File: src/main/java/com/warren/bt/Composite.java
package com.warren.bt;

import java.util.ArrayList;
import java.util.List;

public abstract class Composite implements Node {
  protected final List<Node> children = new ArrayList<>();
  protected int currentChild = 0;

  public Composite addChild(Node child) {
    children.add(child);
    return this;
  }

  public Composite addChildren(Node... nodes) {
    for (Node node : nodes) {
      children.add(node);
    }
    return this;
  }

  @Override
  public void reset() {
    currentChild = 0;
    for (Node child : children) {
      child.reset();
    }
  }

  protected boolean hasChildren() {
    return !children.isEmpty();
  }
}

File: src/main/kotlin/com/warren/bt/Composite.kt
package com.warren.bt

abstract class Composite(override val name: String = "Composite") : Node(name) {
    protected val children = mutableListOf<Node>()

    fun addChild(node: Node) {
        children += node
    }

    override fun reset() {
        children.forEach { it.reset() }
        onReset()
    }

    protected open fun onReset() {}
}
